pkg/resource/deploy/step_generator.go

```diff
@@ -883,13 +858,122 @@ func (sg *stepGenerator) continueStepsFromRefresh(event ContinueResourceRefreshE
+			// but a goroutine blocked on Result and then posting to a channel is very cheap.
+			old, err := cts.Promise().Result(context.Background())
+
+			// Create a fresh 'new' state which will be for the final goal state, the import step will have 'mutated'
+			// the original 'new' variable.
+			var new *resource.State
+			if err == nil {
+				new = resource.NewState(
+					goal.Type, urn, goal.Custom, false, "", goal.Properties, nil, goal.Parent, protectState, false,
+					goal.Dependencies, goal.InitErrors, goal.Provider, goal.PropertyDependencies, false,
+					goal.AdditionalSecretOutputs, aliasUrns, &goal.CustomTimeouts, "", retainOnDelete, goal.DeletedWith,
+					createdAt, modifiedAt, goal.SourcePosition, goal.IgnoreChanges, goal.ReplaceOnChanges)
+			}
+
+			sg.events <- &continueResourceImportEvent{
+				RegisterResourceEvent: event,
+				err:                   err,
+				urn:                   urn,
+				old:                   old,
+				new:                   new,
+				provider:              prov,
+				invalid:               event.Invalid(),
+				recreating:            recreating,
+				randomSeed:            randomSeed,
+			}
+		}()
+
+		sg.imports[urn] = true
 		if isReplace := old != nil && !recreating; isReplace {
 			return []Step{
-				NewImportReplacementStep(sg.deployment, event, old, new, goal.IgnoreChanges, randomSeed),
+				NewImportReplacementStep(sg.deployment, event, old, new, goal.IgnoreChanges, randomSeed, cts),
 				NewReplaceStep(sg.deployment, old, new, nil, nil, nil, true),
-			}, false, nil
+			}, true, nil
+		}
+		return []Step{NewImportStep(sg.deployment, event, new, goal.IgnoreChanges, randomSeed, cts)}, true, nil
+	}
+
+	// Anything else just flow on to the normal step generation.
+	continueEvent := &continueResourceImportEvent{
+		RegisterResourceEvent: event,
+		urn:                   urn,
+		old:                   old,
+		new:                   new,
+		provider:              prov,
+		invalid:               event.Invalid(),
+		recreating:            recreating,
+		randomSeed:            randomSeed,
+	}
+
+	return sg.continueStepsFromImport(continueEvent)
+}
+
+// This function is called by the deployment executor in response to a ContinueResourceImportEvent. It simply
+// calls into continueStepsFromImport and then validateSteps to continue the work that GenerateSteps would
+// have done without an import step.
+func (sg *stepGenerator) ContinueStepsFromImport(event ContinueResourceImportEvent) ([]Step, bool, error) {
+	steps, async, err := sg.continueStepsFromImport(event)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if async {
+		// We only need to validate _real_ steps. If we're returning async work then steps should just be a
+		// DiffStep.
+		return steps, true, nil
+	}
+
+	steps, err = sg.validateSteps(steps)
+	return steps, false, err
+}
+
+func (sg *stepGenerator) continueStepsFromImport(event ContinueResourceImportEvent) ([]Step, bool, error) {
```

Comment:
Think a doc comment about what happens when continuing from an import would be valuable here.

Reply:
Yeah. The naming is pretty confusing here as well given that this also continues from non-import step gen? Maybe `continueStepsWithOld` or something?

Reply:
I suppose it's consistent with our other `continueSteps` functions, at least.

Reply:
ack, I'd suggest we merge as is but I'll solicit bikeshedding for all the continue names to see if there's a better naming for them.